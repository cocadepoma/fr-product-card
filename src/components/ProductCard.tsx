import React, { createContext } from 'react';
import { useProduct } from '../hooks/useProduct';

import { ProductContextProps, Product, onChangeArgs, InitialValues, ProductCardHandlers } from '../interfaces/interfaces';

import styles from '../styles/styles.module.css';

import { ProductImage } from './ProductImage';
import { ProductTitle } from './ProductTitle';
import { ProductButtons } from './ProductButtons';

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export interface Props {
  product: Product;
  // children?: React.ReactElement | React.ReactElement[];
  children: (args: ProductCardHandlers) => JSX.Element;
  className?: string;
  style?: React.CSSProperties;
  onChange?: (args: onChangeArgs) => void;
  value?: number;
  initialValues?: InitialValues;
};

export const ProductCard = ({ children, product, className, style, onChange, value, initialValues }: Props) => {
  const { counter, increaseBy, maxCount, reset, isMaxCountReached } = useProduct({
    onChange,
    product,
    value,
    initialValues,
  });

  return (
    <Provider
      value={{
        counter,
        increaseBy,
        product,
        maxCount,
      }}
    >
      <div
        className={`${styles.productCard} ${className}`}
        style={style}
      >
        {children({
          counter,
          isMaxCountReached,
          maxCount: initialValues?.maxCount,
          increaseBy,
          reset,
          product,
        })}
      </div>
    </Provider>
  )
};

ProductCard.Title = ProductTitle;
ProductCard.Image = ProductImage;
ProductCard.Buttons = ProductButtons;
