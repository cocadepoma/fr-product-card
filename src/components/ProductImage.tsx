import React, { useContext } from "react";
import { ProductContext } from './ProductCard';

import styles from '../styles/styles.module.css';
import noImage from '../assets/no-image.jpg';

export interface Props {
  image?: string;
  className?: string;
  style?: React.CSSProperties;
};

export const ProductImage = ({ image, className, style }: Props) => {
  const { product } = useContext(ProductContext);

  return (
    <img
      className={`${styles.productImg} ${className}`}
      src={image || product.image || noImage}
      alt="Coffee Mug"
      style={style}
    />
  )
};