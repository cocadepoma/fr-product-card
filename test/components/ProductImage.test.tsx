import * as React from 'react';
import renderer from 'react-test-renderer';
import { ProductImage, ProductCard } from '../../src/components';
import { product2 } from '../data/data';

describe('ProductTitle', () => {
  test('should render with the custom image', () => {
    const wrapper = renderer.create(<ProductImage image="https://my-custom-image.png" className="my-custom-class" />);
    expect(wrapper).toMatchSnapshot();
  });

  test('should render with the image of the product', () => {
    const wrapper = renderer.create(
      <ProductCard product={product2}>
        {() => (
          <ProductImage />
        )}
      </ProductCard>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
