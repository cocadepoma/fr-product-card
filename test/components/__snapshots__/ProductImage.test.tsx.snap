// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTitle should render with the custom image 1`] = `
<img
  alt="Coffee Mug"
  className="productImg my-custom-class"
  src="https://my-custom-image.png"
/>
`;

exports[`ProductTitle should render with the image of the product 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Coffee Mug"
    className="productImg undefined"
    src="./coffee-mug.png"
  />
</div>
`;
